File: ADOPCION_I/02_documentacion/ejemplo_antes/sin_documentacion.py
"""
Código ANTES: Sin documentación
Este código funciona pero es difícil de entender sin leerlo línea por línea
"""

def process_payment(amount, currency, method):
    if amount <= 0:
        return None
    
    fee = amount * 0.029 + 0.30
    total = amount + fee
    
    transaction = {
        'amount': amount,
        'fee': fee,
        'total': total,
        'currency': currency,
        'method': method,
        'status': 'pending'
    }
    
    if method == 'card':
        transaction['status'] = 'approved'
    elif method == 'bank':
        transaction['status'] = 'pending_verification'
    
    return transaction


def validate_user_input(data):
    errors = []
    
    if 'email' not in data or '@' not in data['email']:
        errors.append('email')
    
    if 'password' not in data or len(data['password']) < 8:
        errors.append('password')
    
    if 'age' in data and data['age'] < 18:
        errors.append('age')
    
    return len(errors) == 0, errors


class UserManager:
    def __init__(self, db_connection):
        self.db = db_connection
        self.cache = {}
    
    def get_user(self, user_id, use_cache=True):
        if use_cache and user_id in self.cache:
            return self.cache[user_id]
        
        user = self.db.query('SELECT * FROM users WHERE id = ?', [user_id])
        
        if user and use_cache:
            self.cache[user_id] = user
        
        return user
    
    def update_user(self, user_id, data):
        result = self.db.execute(
            'UPDATE users SET name=?, email=? WHERE id=?',
            [data.get('name'), data.get('email'), user_id]
        )
        
        if user_id in self.cache:
            del self.cache[user_id]
        
        return result

File: ADOPCION_I/02_documentacion/ejemplo_antes/sin_documentacion.js
/**
 * Código ANTES: Sin documentación
 * Este código funciona pero es difícil de entender sin leerlo línea por línea
 */

function processPayment(amount, currency, method) {
    if (amount <= 0) {
        return null;
    }
    
    const fee = amount * 0.029 + 0.30;
    const total = amount + fee;
    
    const transaction = {
        amount: amount,
        fee: fee,
        total: total,
        currency: currency,
        method: method,
        status: 'pending'
    };
    
    if (method === 'card') {
        transaction.status = 'approved';
    } else if (method === 'bank') {
        transaction.status = 'pending_verification';
    }
    
    return transaction;
}

function validateUserInput(data) {
    const errors = [];
    
    if (!data.email || !data.email.includes('@')) {
        errors.push('email');
    }
    
    if (!data.password || data.password.length < 8) {
        errors.push('password');
    }
    
    if (data.age && data.age < 18) {
        errors.push('age');
    }
    
    return {
        isValid: errors.length === 0,
        errors: errors
    };
}

class UserManager {
    constructor(dbConnection) {
        this.db = dbConnection;
        this.cache = {};
    }
    
    getUser(userId, useCache = true) {
        if (useCache && this.cache[userId]) {
            return this.cache[userId];
        }
        
        const user = this.db.query('SELECT * FROM users WHERE id = ?', [userId]);
        
        if (user && useCache) {
            this.cache[userId] = user;
        }
        
        return user;
    }
    
    updateUser(userId, data) {
        const result = this.db.execute(
            'UPDATE users SET name=?, email=? WHERE id=?',
            [data.name, data.email, userId]
        );
        
        if (this.cache[userId]) {
            delete this.cache[userId];
        }
        
        return result;
    }
}

module.exports = { processPayment, validateUserInput, UserManager };
